contrib/postgres_inspector/filters/listener/source/postgres_inspector.cc

```diff
@@ -0,0 +1,280 @@
+#include "contrib/postgres_inspector/filters/listener/source/postgres_inspector.h"
+
+#include "envoy/network/listen_socket.h"
+#include "envoy/stats/scope.h"
+
+#include "source/common/api/os_sys_calls_impl.h"
+#include "source/common/buffer/buffer_impl.h"
+#include "source/common/common/assert.h"
+#include "source/common/common/empty_string.h"
+#include "source/common/network/address_impl.h"
+#include "source/common/protobuf/protobuf.h"
+#include "source/common/protobuf/utility.h"
+
+#include "contrib/envoy/extensions/filters/listener/postgres_inspector/v3alpha/postgres_inspector.pb.h"
+
+namespace Envoy {
+namespace Extensions {
+namespace ListenerFilters {
+namespace PostgresInspector {
+
+Config::Config(Stats::Scope& scope, const ProtoConfig& proto_config)
+    : stats_(PostgresInspectorStats{
+          ALL_POSTGRES_INSPECTOR_STATS(POOL_COUNTER_PREFIX(scope, "postgres_inspector"),
+                                       POOL_HISTOGRAM_PREFIX(scope, "postgres_inspector"))}),
+      enable_metadata_extraction_(
+          PROTOBUF_GET_WRAPPED_OR_DEFAULT(proto_config, enable_metadata_extraction, true)),
+      max_startup_message_size_(PROTOBUF_GET_WRAPPED_OR_DEFAULT(
+          proto_config, max_startup_message_size, DEFAULT_MAX_STARTUP_MESSAGE_SIZE)),
+      startup_timeout_(std::chrono::milliseconds(
+          PROTOBUF_GET_MS_OR_DEFAULT(proto_config, startup_timeout, DEFAULT_STARTUP_TIMEOUT_MS))) {}
+
+Filter::Filter(const ConfigSharedPtr& config) : config_(config) {}
+
+Network::FilterStatus Filter::onAccept(Network::ListenerFilterCallbacks& cb) {
+  ENVOY_LOG(trace, "postgres inspector: new connection accepted");
+  cb_ = &cb;
+
+  // Set up timeout timer.
+  if (config_->startupTimeout().count() > 0) {
+    timeout_timer_ = cb.dispatcher().createTimer([this]() { onTimeout(); });
+    timeout_timer_->enableTimer(config_->startupTimeout());
+  }
+
+  return Network::FilterStatus::StopIteration;
+}
+
+Network::FilterStatus Filter::onData(Network::ListenerFilterBuffer& buffer) {
+  const auto raw_slice = buffer.rawSlice();
+  ENVOY_LOG(debug, "postgres inspector: onData called with {} bytes, bytes_read_={}, state_={}",
+            raw_slice.len_, bytes_read_, static_cast<int>(state_));
+
+  // Skip data we've already processed.
+  if (static_cast<uint64_t>(raw_slice.len_) <= bytes_read_) {
+    ENVOY_LOG(debug, "postgres inspector: skipping already processed data ({} <= {})",
+              raw_slice.len_, bytes_read_);
+    return Network::FilterStatus::StopIteration;
+  }
+
+  // Update bytes read after processing state logic to account for drains.
+  switch (state_) {
+  case ParseState::Initial: {
+    auto status = processInitialData(buffer);
+    // After processInitialData we may have drained 8 bytes. We recompute bytes_read_.
+    bytes_read_ = buffer.rawSlice().len_;
+    return status;
+  }
+  case ParseState::ProcessingStartup: {
```

Comment:
I don't think this `case` will be trigger at all because if the message is not SSL request, you call `processStartupMessage` at the end of `processInitialData`. 
We won't have a state where `onData` is called with state set to `ProcessingStartup`.
Even if the client sends the startup message as the first message because you will always called `processInitialData` and that will call `processStartupMessage` in the end.